middle_end/flambda2/types/meet_and_join.ml

```diff
@@ -842,9 +842,46 @@ and meet_product_value_slot_indexed env
     ({ value_slot_components_by_index = components_by_index2 } :
       TG.Product.Value_slot_indexed.t) :
     (TG.Product.Value_slot_indexed.t * TEE.t) Or_bottom.t =
+  let components_by_index1' =
+    Value_slot.Map.fold
+      (fun value_slot (ty, _) components_by_index ->
+        Value_slot.Map.add value_slot ty components_by_index)
+      components_by_index1 Value_slot.Map.empty
+  in
+  let components_by_index2' =
+    Value_slot.Map.fold
+      (fun value_slot (ty, _) components_by_index ->
+        Value_slot.Map.add value_slot ty components_by_index)
+      components_by_index2 Value_slot.Map.empty
+  in
   let<+ components_by_index, env_extension =
-    meet_generic_product env ~components_by_index1 ~components_by_index2
-      ~union:Value_slot.Map.union
+    meet_generic_product env ~components_by_index1:components_by_index1'
+      ~components_by_index2:components_by_index2' ~union:Value_slot.Map.union
```

Comment:
I think that instead of splitting off the kinds, you could have just defined a custom union:
```ocaml
let union f m1 m2 =
  Value_slot.Map.union
    (fun value_slot (ty1, kind1) (ty2, kind2) ->
       if not (K.with_subkind.equal kind1 kind2) then Misc.fatal_errorf ...;
       Option.map (fun ty -> ty, kind1) (f value_slot ty1 ty2))
    m1 m2
in ...
```